### PR TITLE
[Firo]: Support exchange address

### PR DESCRIPTION
**Pull request by w20089527**
## Description

A new address for exchange requirement.

As exchange address has same unlocking script with regular address(p2pkh), our work will involve as following:
1. Support `derive` of this address
2. Support `validation` of this address
3. Support this address as an output which added a new instruction `OP_EXCHANGEADDR`

FYI: as we do not have any config in `registry.json` for this kind of address, I defined a `ExchangeAddress` with constants that only works for `mainnet`, if want to support `testnet` or `regtest`, we need to define corresponding class for them.

Check here for exchange address description:
https://github.com/firoorg/firo/wiki/Exchange-Addresses

Here is the detail pr for exchange address:
https://github.com/firoorg/firo/pull/1356

## How to test
Testcases added for this feature

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist

- [ ] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [ ] Update documentation as needed.
- [ ] If there is a related Issue, mention it in the description.

If you're adding a new blockchain

- [ ] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

### CODE REVIEW

File: tests/chains/Firo/TWZCoinAddressTests.cpp [hunk 1]
```diff
@@ -22,6 +24,23 @@ TEST(TWZCoin, Address) {
...
+    auto exchangeAddressData = WRAPD(TWAnyAddressData(exchangeAddress.get()));
+    assertHexEqual(exchangeAddressData, "c7529bf17541410428c7b23b402761acb83fdfba");
+}
+
+TEST(TWZCoin, ExchangeAddress_DeriveFromPublicKey) {
+    auto publicKey = WRAP(TWPublicKey, TWPublicKeyCreateWithData(DATA("034cc1963365aa67d35643f419d6601eca6ef7f62e46bf7f8b6ffa64e2f44fd0bf").get(), TWPublicKeyTypeSECP256k1));
+    auto address = WRAP(TWAnyAddress, TWAnyAddressCreateWithPublicKeyDerivation(publicKey.get(), TWCoinTypeFiro, TWDerivationFiroExchange));
+    auto addressDesc = WRAPS(TWAnyAddressDescription(address.get()));
+    assertStringsEqual(addressDesc, "EXXWKhUtcaFKVW1NeRFuqPq33zAJMtQJwR4y");
+}
```
Comment (satoshiotomakan) [inline tests/chains/Firo/TWZCoinAddressTests.cpp:42]: Please consider moving these tests to `TWFiroAddressTests.cpp`

File: registry.json [hunk 2]
```diff
@@ -1363,6 +1363,12 @@
         "path": "m/44'/136'/0'/0/0",
         "xpub": "xpub",
         "xprv": "xprv"
+      },
+      {
+        "name": "exchange",
+        "path": "m/44'/136'/0'/0/0",
+        "xpub": "xpub",
+        "xprv": "xprv"
       }
```
Comment (satoshiotomakan) [inline registry.json:1372]: We do not add derivations for such purposes. Instead, could you consider adding new `Prefix` variant like for Filecoin delegated address:
https://github.com/trustwallet/wallet-core/blob/c0cf03fe36c0b3116b7913bcb199cf3017f966b1/src/CoinEntry.h#L34-L37
FFI:
https://github.com/trustwallet/wallet-core/blob/c0cf03fe36c0b3116b7913bcb199cf3017f966b1/include/TrustWalletCore/TWAnyAddress.h#L117-L123
Usage:
https://github.com/trustwallet/wallet-core/blob/c0cf03fe36c0b3116b7913bcb199cf3017f966b1/src/Filecoin/Entry.cpp#L19-L24

**Review by satoshiotomakan — CHANGES_REQUESTED** [review]
Hi @w20089527, thank you for opening the PR!

File: src/CoinEntry.h [hunk 3]
```diff
@@ -34,7 +34,10 @@ using SS58Prefix = uint32_t;
 /// Declare a dummy prefix to notify the entry to derive a delegated address.
 struct DelegatedPrefix {};
 
-using PrefixVariant = std::variant<Base58Prefix, Bech32Prefix, SS58Prefix, DelegatedPrefix, std::monostate>;
+/// Declare a dummy prefix to notify the entry to derive a firo exchange address.
+struct ExchangePrefix {};
```
Comment (satoshiotomakan) [inline src/CoinEntry.h:38]: If it's necessary to support testnet exchange addresses, this `ExchangePrefix` structure could include `p2pkh` prefix (or also `p2sh` if applied).
What do you think?

File: src/interface/TWAnyAddress.cpp [hunk 4]
```diff
@@ -90,6 +90,14 @@ struct TWAnyAddress* TWAnyAddressCreateWithPublicKeyFilecoinAddressType(struct T
     return new TWAnyAddress{TW::AnyAddress::createAddress(publicKey->impl, TWCoinTypeFilecoin, TWDerivationDefault, prefix)};
 }
 
+struct TWAnyAddress* TWAnyAddressCreateWithPublicKeyFiroAddressType(struct TWPublicKey* _Nonnull publicKey, enum TWFiroAddressType firoAddressType) {
+    TW::PrefixVariant prefix = std::monostate();
+    if (firoAddressType == TWFiroAddressTypeExchange) {
+        prefix = TW::ExchangePrefix();
```
Comment (satoshiotomakan) [inline src/interface/TWAnyAddress.cpp:96]: In addition to the comment above, this function could take p2pkh and(?) p2sh prefixes

**Review by satoshiotomakan — APPROVED** [review]
Thank you for the fix! LGTM
I have one suggestion in case if you need to support testnet addresses

File: src/CoinEntry.h (hunk 3 above)
Comment (w20089527) [reply src/CoinEntry.h:38]: Thanks for the review. Actually, we do not have to support testnet now, if any requirement, we can extend this structure to support it. So I prefer keeping it empty to return mainnet address.

File: src/interface/TWAnyAddress.cpp (hunk 4 above)
Comment (w20089527) [reply src/interface/TWAnyAddress.cpp:96]: Actually, it is a kind of p2pkh address, it is only different between mainnet and testnet. If want to support testnet, we can define a testnet address type and make some change in `ExchangeAddress` for it.

**Review by Milerius — APPROVED** [review]

**Pull request merged**